logic/resolvable/Concludable.java

```diff
@@ -268,67 +267,61 @@ public Relation(final RelationConstraint constraint, @Nullable IsaConstraint isa
-                    .filter(thenIdx -> unificationSatisfiable(conjRP, thenRolePlayers.get(thenIdx), conceptMgr))
-                    .map(thenIdx -> {
-                        Map<RolePlayer, Set<Integer>> clone = cloneMapping(mapping);
+            Set<RolePlayer> thenRolePlayers = relationConclusion.relation().players();
+
+            return matchRolePlayers(conjRolePlayers, thenRolePlayers, new HashMap<>(), conceptMgr)
+                    .map(mapping -> convertRPMappingToUnifier(mapping, unifierBuilder.duplicate(), conceptMgr));
+        }
+
+        private ResourceIterator<Map<RolePlayer, Set<RolePlayer>>> matchRolePlayers(
+                List<RolePlayer> conjRolePLayers, Set<RolePlayer> thenRolePlayers,
+                Map<RolePlayer, Set<RolePlayer>> mapping, ConceptManager conceptMgr) {
+            if (conjRolePLayers.isEmpty()) return Iterators.iterate(list(mapping));
+            RolePlayer conjRP = conjRolePLayers.get(0);
+            return Iterators.iterate(thenRolePlayers)
+                    .filter(thenRP -> Iterators.iterate(mapping.values()).noneMatch(rolePlayers -> rolePlayers.contains(thenRP)))
+                    .filter(thenRP -> unificationSatisfiable(conjRP, thenRP, conceptMgr))
+                    .map(thenRP -> {
+                        Map<RolePlayer, Set<RolePlayer>> clone = cloneMapping(mapping);
                         clone.putIfAbsent(conjRP, new HashSet<>());
-                        clone.get(conjRP).add(thenIdx);
+                        clone.get(conjRP).add(thenRP);
                         return clone;
-                    }).flatMap(newMapping -> matchRolePlayerIndices(conjRolePlayers.subList(1, conjRolePlayers.size()),
-                                                                    thenRolePlayers, newMapping, conceptMgr));
+                    }).flatMap(newMapping -> matchRolePlayers(conjRolePLayers.subList(1, conjRolePLayers.size()),
+                                                              thenRolePlayers, newMapping, conceptMgr));
+        }
+
+        private Unifier convertRPMappingToUnifier(
+                Map<RolePlayer, Set<RolePlayer>> mapping, Unifier.Builder unifierBuilder, ConceptManager conceptMgr) {
+            mapping.forEach((conjRP, thenRPs) -> thenRPs.forEach(thenRP -> {
+                unifierBuilder.add(conjRP.player().id(), thenRP.player().id());
+                if (conjRP.roleType().isPresent()) {
+                    assert thenRP.roleType().isPresent();
+                    TypeVariable conjRoleType = conjRP.roleType().get();
+                    TypeVariable thenRoleType = thenRP.roleType().get();
+                    unifierBuilder.add(conjRoleType.id(), thenRoleType.id());
+                    if (conjRoleType.reference().isLabel()) {
+                        Set<Label> allowedTypes = conjRoleType.resolvedTypes().stream()
+                                .flatMap(roleLabel -> subtypeLabels(roleLabel, conceptMgr))
```

Comment:
Kept as `.stream()` because subtypeRoleLabels returns a `Stream`. I suspect this all needs to be refactored to use Grakn `Iterator`, but for now this I'll keep using `.stream()`.